C/services/core/configuration_manager.cpp

```diff
@@ -0,0 +1,512 @@
+ * Author: Massimiliano Pinto
+ */
+
+#include <configuration_manager.h>
+#include <rapidjson/writer.h>
+
+using namespace std;
+using namespace rapidjson;
+
+ConfigurationManager *ConfigurationManager::m_instance = 0;
+
+/**
+ * Constructor
+ *
+ * @param host    Storage layer TCP address
+ * @param port	  Storage layer TCP port
+ */
+ConfigurationManager::ConfigurationManager(const string& host,
+					   unsigned short port)
+{
+	m_storage = new StorageClient(host, port);
+}
+
+// Destructor
+ConfigurationManager::~ConfigurationManager()
+{
+	delete m_storage;
+}
+
+/**
+ * Return the singleton instance of the configuration manager
+ *
+ * @param host    Storage layer TCP address
+ * @param port	  Storage layer TCP port
+ * @return        The configuration manager class instance
+ */
+ConfigurationManager* ConfigurationManager::getInstance(const string& host,
+							unsigned short port)
+{
+	if (m_instance == 0)
+	{
+		m_instance = new ConfigurationManager(host, port);
+	}
+	return m_instance;
+}
+
+/**
+ * Return all FogLAMP categories from storage layer
+ *
+ * @return	ConfigCategories class object with
+ *		key and description for all found categories.
+ */
+ConfigCategories ConfigurationManager::getAllCategoryNames() const
+{
+	vector<Returns *> columns;
+	columns.push_back(new Returns("key"));
+	columns.push_back(new Returns("description"));
+	Query qAllCategories(columns);
+
+	// Query via Storage client
+	ResultSet* allCategories = m_storage->queryTable("configuration", qAllCategories);
+
+	ConfigCategories categories;
+
+	for (ResultSet::RowIterator it = allCategories->firstRow(); ;)
+	{
+		ResultSet::Row* row = *it;
+		if (!row)
+		{
+			delete allCategories;
+			throw NoSuchRowxception();
+		}	
+		ResultSet::ColumnValue* key = row->getColumn("key");
+		ResultSet::ColumnValue* description = row->getColumn("description");
+
+		ConfigCategoryDescription *value = new ConfigCategoryDescription(key->getString(),
+										 description->getString());
+
+		// Add current row data to categories;
+		categories.addCategoryDescription(value);
+	
+		if (allCategories->isLastRow(it))
+		{
+			break;
+		}
+
+		it++;
+	}
+
+	delete allCategories;
+
+	return categories;
+}
+
+/**
+ * Return all the items of a specific category
+ * from the storage layer.
+ *
+ * @param categoryName	The specified category name
+ * @return		ConfigCategory calss object
+ *			with all category items
+ * @throw 		NoSuchCategoryException
+ * @throw		Exception
+ */
+
+ConfigCategory ConfigurationManager::getCategoryAllItems(const string& categoryName) const
+{
+	// SELECT * FROM foglamp.configuration WHERE key = categoryName
+	const Condition conditionKey(Equals);
+	Where *wKey = new Where("key", conditionKey, categoryName);
+	Query qKey(wKey);
+
+	// Query via storage client
+	ResultSet* categoryItems = m_storage->queryTable("configuration", qKey);
+
+	// Cayegory not found
+	if (!categoryItems->rowCount())
+	{
+		delete categoryItems;
+		throw NoSuchCategoryException();
```

Comment:
Given the call is about categories throwing an exception about rows is a little odd. I would have thought a category based exception would be better. Also the naming of this reception is odd, I guess it is because it conflicts with another exception.